gs_quant/timeseries/arima.py

```diff
@@ -0,0 +1,248 @@
+        error = mse(test, yhat)
+
+        return error, const, ar_coef, ma_coef, resid
+
+
+    def fit(self, X: Union[pd.Series, pd.DataFrame], train_size: float, p_vals: list = [0,1,2], d_vals: list=[0,1,2], q_vals: list=[0,1,2], freq: str=None) -> arima: 
+        """
+        Train a combination of ARIMA models. If pandas DataFrame, finds the 
+        best arima model parameters for each column. If pandas Series, finds 
+        the best arima model parameters for the series.
+
+        :param X: time series to be operated on; required parameter
+        :param train_size: between 0.0 and 1.0 and represents the proportion of the dataset to include in the train split
+        :p_vals: number of autoregressive terms to search; default is [0,1,2]
+        :d_vals: number of differences to search; default is [0,1,2]
+        :q_vals: number of lagged forecast to search; always [0,1,2]
+        :freq: frequency of time series, default is None
+        :return: self
+        """
+
+        if isinstance(X, pd.Series): X = X.to_frame()
+
+        if isinstance(X, pd.DataFrame):
+            for series_id in tqdm(X.columns):
+                series = X[series_id]
+                best_score, best_order, best_const, best_ar_coef, best_ma_coef, best_resid = float("inf"), None, None, None, None, None
+                for order in list(itertools.product(*[p_vals, d_vals, q_vals])):
+                    try:
+                        mse, const, ar_coef, ma_coef, resid = self._evaluate_arima_model(series, order, train_size, freq)
+                        if mse < best_score:
+                            best_score = mse
+                            best_order = order
+                            best_const = const
+                            best_ar_coef = ar_coef
+                            best_ma_coef = ma_coef
+                            best_resid = resid
+                    except Exception as e:
+                        print('   {}'.format(e))
+                        continue
+
+                p, d, q = best_order
+                assert(p == len(best_ar_coef))
+
+                self.best_params[series_id] = ARIMA_BestParams(freq=freq,
+                                                               p=p, 
+                                                               d=d, 
+                                                               q=q,
+                                                               const=best_const,
+                                                               ar_coef=best_ar_coef,
+                                                               ma_coef=best_ma_coef,
+                                                               resid=best_resid,
+                                                               series=series)
+        else:
+            raise ValueError("Not DataFrame or Series!")
+        return self 
+    
+
+    # Helper Function to Difference Time Series n Times
```

Comment:
any function descriptions pls add as a header ("""this format directly under signature""") - pls look at rest of lib for examples